### PR TITLE
More thread-safety changes

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister

### CODE REVIEW

**Comment by michaelwoerister** [comment]
@rust-lang/compiler Does anybody know what `LazyTokenStream` is? It seems to be unused.

**Comment by petrochenkov** [comment]
@michaelwoerister 
>Does anybody know what LazyTokenStream is?

It's a hack! But it's used.
Many nonterminal tokens (e.g. `NtExpr`) don't have their original token streams from which they were parsed (they have only AST fragments like `ast::Expr`),  but these token streams are sometimes needed (for example if they are a part of macro definition and that macro definition needs to be stored in metadata), so AST fragments are pretty-printed into text (!) and then parsed back into the token stream.
`LazyTokenStream` serves as a cache for these AST -> TokenStream conversions to avoid doing these conversions multiple times, so removing it won't affect correctness, only performance.

**Comment by Zoxc** [comment]
@petrochenkov Why do we keep the AST around in the token stream though? It seems more correct to use the original TokenStream instead of relying on an lossy AST -> TokenStream transformation

**Comment by petrochenkov** [comment]
@Zoxc 
>Why do we keep the AST around in the token stream though?

Who knows. `Nonterminal` is represented with AST fragments since at least 2012, years before token streams were introduced.
Someone needs to port expansion code working with nonterminals to token trees.

**Comment by shepmaster** [comment]
Ping from triage, @michaelwoerister !

**Comment by michaelwoerister** [comment]
@Zoxc, looks good to me except for `intern_inlined_body`. I think you can just remove the assertion there.

File: src/librustc/hir/map/mod.rs [hunk 1]
```diff
@@ -927,8 +928,13 @@ impl<'hir> Map<'hir> {
     }
 
     pub fn intern_inlined_body(&self, def_id: DefId, body: Body) -> &'hir Body {
+        let mut inlined_bodies = self.inlined_bodies.borrow_mut();
+        if let Some(&b) = inlined_bodies.get(&def_id) {
+            debug_assert_eq!(&body, b);
```
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc/hir/map/mod.rs]: This isn't a pointer comparison. This check ensures that we always give the same body for the same def_id to intern_inlined_body.

[the remark this answers is not in the document]
Comment (michaelwoerister) [reply src/librustc/hir/map/mod.rs]: Oh, you're right of course `:)`

**Pull request merged**